virtual_ecosystem/models/litter/chemistry.py

```diff
@@ -0,0 +1,530 @@
+impact of N and P stoichiometry on litter decomposition rates. By contrast, the impact
+of lignin on decay rates is directly calculated.
+"""  # noqa: D205
+
+import numpy as np
+from numpy.typing import NDArray
+from xarray import DataArray
+
+from virtual_ecosystem.core.data import Data
+from virtual_ecosystem.models.litter.constants import LitterConsts
+
+
+class LitterChemistry:
+    """This class handles the chemistry of litter pools.
+
+    This class contains methods to calculate the changes in the litter pool chemistry
+    based on the contents of the `data` object, as well as method to calculate total
+    mineralisation based on litter pool decay rates.
+    """
+
+    def __init__(self, data: Data, constants: LitterConsts):
+        self.data = data
+        self.structural_to_metabolic_n_ratio = constants.structural_to_metabolic_n_ratio
+
+    def calculate_new_pool_chemistries(
+        self,
+        plant_inputs: dict[str, NDArray[np.float32]],
+        metabolic_splits: dict[str, NDArray[np.float32]],
+        updated_pools: dict[str, NDArray[np.float32]],
+    ) -> dict[str, DataArray]:
+        """Method to calculate the updated chemistry of each litter pool.
+
+        All pools contain nitrogen and phosphorus, so this is updated for every pool.
+        Only the structural (above and below ground) pools and the woody pools contain
+        lignin, so it is only updated for those pools.
+
+        Args:
+            plant_inputs: Dictionary containing the amount of each litter type that is
+                added from the plant model in this time step [kg C m^-2]
+            metabolic_splits: Dictionary containing the proportion of each input that
+                goes to the relevant metabolic pool. This is for three input types:
+                leaves, reproductive tissues and roots [unitless]
+            updated_pools: Dictionary containing the updated pool densities for all 5
+                litter pools [kg C m^-2]
+        """
+
+        # Find lignin and nitrogen contents of the litter input flows
+        input_lignin = self.calculate_litter_input_lignin_concentrations(
+            plant_input_above_struct=plant_inputs["above_ground_structural"],
+            plant_input_below_struct=plant_inputs["below_ground_structural"],
+        )
+        input_c_n_ratios = self.calculate_litter_input_nitrogen_ratios(
+            metabolic_splits=metabolic_splits,
+            struct_to_meta_nitrogen_ratio=self.structural_to_metabolic_n_ratio,
+        )
+
+        # Then use to find
+        change_in_lignin = self.calculate_lignin_updates(
+            plant_inputs=plant_inputs,
+            input_lignin=input_lignin,
+            updated_pools=updated_pools,
+        )
+        change_in_c_n_ratios = self.calculate_c_n_ratio_updates(
+            plant_inputs=plant_inputs,
+            input_c_n_ratios=input_c_n_ratios,
+            updated_pools=updated_pools,
+        )
+
+        return {
```

Comment:
This could probably be condensed in a something that goes over a list

Reply:
Ahh yes I was wondering about whether I should do that!